File: conversions/uuid/elide-uuid.ts
/**
 * UUID - Universally Unique Identifier Generator
 *
 * Generate RFC 4122 compliant UUIDs (v4).
 * **POLYGLOT SHOWCASE**: One UUID generator for ALL languages on Elide!
 *
 * Features:
 * - Generate UUIDv4 (random)
 * - Validate UUID format
 * - Parse UUID components
 * - NIL UUID support
 * - RFC 4122 compliant
 *
 * Polyglot Benefits:
 * - Python, Ruby, Java all need UUID generation
 * - ONE implementation works everywhere on Elide
 * - Consistent IDs across languages
 * - No need for language-specific UUID libs
 *
 * Use cases:
 * - Database primary keys
 * - API request IDs
 * - Session identifiers
 * - File naming
 * - Distributed systems
 * - Message queues
 *
 * Package has ~25M+ downloads/week on npm!
 */

/**
 * NIL UUID (all zeros)
 */
export const NIL = '00000000-0000-0000-0000-000000000000';

/**
 * Generate a random hex digit
 */
function randomHex(): string {
  return Math.floor(Math.random() * 16).toString(16);
}

/**
 * Generate a UUIDv4 (random)
 */
export function v4(): string {
  const hex = Array.from({ length: 32 }, () => randomHex()).join('');

  return [
    hex.substring(0, 8),
    hex.substring(8, 12),
    '4' + hex.substring(13, 16),  // Version 4
    ((parseInt(hex[16], 16) & 0x3) | 0x8).toString(16) + hex.substring(17, 20),  // Variant 10
    hex.substring(20, 32)
  ].join('-');
}

/**
 * Validate UUID format
 */
export function validate(uuid: string): boolean {
  if (!uuid || typeof uuid !== 'string') {
    return false;
  }

  const pattern = /^[0-9a-f]{8}-[0-9a-f]{4}-[1-5][0-9a-f]{3}-[89ab][0-9a-f]{3}-[0-9a-f]{12}$/i;
  return pattern.test(uuid);
}

/**
 * Parse UUID into components
 */
export function parse(uuid: string): {
  timeLow: string;
  timeMid: string;
  timeHiAndVersion: string;
  clockSeqHiAndReserved: string;
  clockSeqLow: string;
  node: string;
  version: number;
  variant: string;
} | null {
  if (!validate(uuid)) {
    return null;
  }

  const parts = uuid.split('-');
  const version = parseInt(parts[2][0], 16);

  // Determine variant
  const variantBits = parseInt(parts[3][0], 16);
  let variant = 'unknown';
  if ((variantBits & 0x8) === 0) {
    variant = 'NCS';
  } else if ((variantBits & 0xc) === 0x8) {
    variant = 'RFC4122';
  } else if ((variantBits & 0xe) === 0xc) {
    variant = 'Microsoft';
  } else {
    variant = 'Reserved';
  }

  return {
    timeLow: parts[0],
    timeMid: parts[1],
    timeHiAndVersion: parts[2],
    clockSeqHiAndReserved: parts[3].substring(0, 2),
    clockSeqLow: parts[3].substring(2),
    node: parts[4],
    version,
    variant
  };
}

/**
 * Get UUID version
 */
export function version(uuid: string): number | null {
  if (!validate(uuid)) {
    return null;
  }

  return parseInt(uuid.split('-')[2][0], 16);
}

/**
 * Check if UUID is NIL
 */
export function isNil(uuid: string): boolean {
  return uuid === NIL;
}

/**
 * Generate multiple UUIDs
 */
export function generate(count: number): string[] {
  return Array.from({ length: count }, () => v4());
}

// Default export
export default {
  v4,
  validate,
  parse,
  version,
  isNil,
  generate,
  NIL
};

// CLI Demo
if (import.meta.url.includes("elide-uuid.ts")) {
  console.log("🆔 UUID - Universal Identifier Generator for Elide (POLYGLOT!)\\n");

  console.log("=== Example 1: Generate UUIDs ===");
  console.log("Single UUID:", v4());
  console.log("Another UUID:", v4());
  console.log("One more:", v4());
  console.log();

  console.log("=== Example 2: UUID Format ===");
  const uuid1 = v4();
  console.log("UUID:", uuid1);
  console.log("Format: xxxxxxxx-xxxx-4xxx-yxxx-xxxxxxxxxxxx");
  console.log("  Time Low: 8 hex digits");
  console.log("  Time Mid: 4 hex digits");
  console.log("  Version: 4 (random)");
  console.log("  Variant: RFC 4122");
  console.log("  Node: 12 hex digits");
  console.log();

  console.log("=== Example 3: Validate UUIDs ===");
  const validTests = [
    v4(),
    "123e4567-e89b-12d3-a456-426614174000",
    "invalid-uuid",
    "123e4567-e89b-12d3-a456",
    NIL,
    ""
  ];

  validTests.forEach(uuid => {
    console.log(`  ${uuid || '(empty)'}: ${validate(uuid)}`);
  });
  console.log();

  console.log("=== Example 4: Parse UUID ===");
  const uuid2 = v4();
  console.log("UUID:", uuid2);
  const parsed = parse(uuid2);
  if (parsed) {
    console.log("Parsed:");
    console.log(`  Time Low: ${parsed.timeLow}`);
    console.log(`  Time Mid: ${parsed.timeMid}`);
    console.log(`  Version: ${parsed.version}`);
    console.log(`  Variant: ${parsed.variant}`);
    console.log(`  Node: ${parsed.node}`);
  }
  console.log();

  console.log("=== Example 5: UUID Version ===");
  const uuid3 = v4();
  console.log("UUID:", uuid3);
  console.log("Version:", version(uuid3));
  console.log();

  console.log("=== Example 6: NIL UUID ===");
  console.log("NIL UUID:", NIL);
  console.log("Is valid?", validate(NIL));
  console.log("Is NIL?", isNil(NIL));
  console.log("Random is NIL?", isNil(v4()));
  console.log();

  console.log("=== Example 7: Generate Multiple ===");
  const batch = generate(5);
  console.log("Generated 5 UUIDs:");
  batch.forEach((id, i) => {
    console.log(`  ${i + 1}. ${id}`);
  });
  console.log();

  console.log("=== Example 8: Database IDs ===");
  console.log("User IDs:");
  console.log(`  user-${v4()}`);
  console.log(`  user-${v4()}`);
  console.log(`  user-${v4()}`);
  console.log();

  console.log("=== Example 9: API Request IDs ===");
  console.log("Request tracking:");
  const requestId = v4();
  console.log(`  Request ID: ${requestId}`);
  console.log(`  X-Request-ID: ${requestId}`);
  console.log();

  console.log("=== Example 10: Session IDs ===");
  console.log("Session identifiers:");
  console.log(`  session_${v4()}`);
  console.log(`  session_${v4()}`);
  console.log();

  console.log("=== Example 11: File Names ===");
  const fileExtensions = ['jpg', 'pdf', 'docx', 'mp4'];
  console.log("Unique file names:");
  fileExtensions.forEach(ext => {
    console.log(`  ${v4()}.${ext}`);
  });
  console.log();

  console.log("=== Example 12: Message Queue IDs ===");
  console.log("Message IDs:");
  for (let i = 0; i < 3; i++) {
    const msgId = v4();
    console.log(`  msg-${msgId}`);
  }
  console.log();

  console.log("=== Example 13: Distributed System ===");
  console.log("Transaction IDs across services:");
  console.log(`  Service A: txn-${v4()}`);
  console.log(`  Service B: txn-${v4()}`);
  console.log(`  Service C: txn-${v4()}`);
  console.log();

  console.log("=== Example 14: UUID Uniqueness ===");
  const uuids = generate(1000);
  const unique = new Set(uuids);
  console.log(`Generated: ${uuids.length}`);
  console.log(`Unique: ${unique.size}`);
  console.log(`Collisions: ${uuids.length - unique.size}`);
  console.log("(Should be 0 collisions)");
  console.log();

  console.log("=== Example 15: POLYGLOT Use Case ===");
  console.log("🌐 Same UUID generator works in:");
  console.log("  • JavaScript/TypeScript");
  console.log("  • Python (via Elide)");
  console.log("  • Ruby (via Elide)");
  console.log("  • Java (via Elide)");
  console.log();
  console.log("Benefits:");
  console.log("  ✓ One implementation, all languages");
  console.log("  ✓ Consistent IDs everywhere");
  console.log("  ✓ No language-specific UUID bugs");
  console.log("  ✓ Share ID generation across polyglot projects");
  console.log();

  console.log("✅ Use Cases:");
  console.log("- Database primary keys");
  console.log("- API request IDs");
  console.log("- Session identifiers");
  console.log("- File naming");
  console.log("- Distributed systems");
  console.log("- Message queues");
  console.log();

  console.log("🚀 Performance:");
  console.log("- Zero dependencies");
  console.log("- Instant execution on Elide");
  console.log("- 10x faster than Node.js cold start");
  console.log("- ~25M+ downloads/week on npm");
  console.log();

  console.log("💡 Polyglot Tips:");
  console.log("- Use in Python/Ruby/Java via Elide");
  console.log("- Share ID generation across languages");
  console.log("- One UUID standard for all services");
  console.log("- Perfect for distributed systems!");
}

File: conversions/uuid/elide-uuid.rb
#!/usr/bin/env ruby

# Ruby Integration Example for elide-uuid
#
# This demonstrates calling the TypeScript uuid implementation from Ruby
# using Elide's polyglot capabilities.
#
# Benefits:
# - One UUID implementation shared across TypeScript and Ruby
# - Consistent ID generation across microservices
# - No Ruby UUID gem needed
# - Perfect for Rails, Sidekiq, background workers

puts "=== Ruby Calling TypeScript UUID ===\n"

# NOTE: Exact syntax depends on Elide's Ruby polyglot API (currently alpha)
# This is a CONCEPTUAL example - adjust when Ruby support is ready

# Assuming Elide provides something like:
# uuid_module = Elide.require('./elide-uuid.ts')

# Example 1: Generate UUID
# uuid1 = uuid_module.v4()
# puts "Generated UUID: #{uuid1}"
# puts "Validation: #{uuid_module.validate(uuid1)}"
# puts ""

# Example 2: Use in Rails Model
# class User < ApplicationRecord
#   before_create :generate_uuid
#
#   private
#   def generate_uuid
#     self.uuid = uuid_module.v4()
#   end
# end
#
# user = User.create(name: "Alice")
# puts "User UUID: #{user.uuid}"
# puts ""

# Example 3: Background Job with Sidekiq
# class DataProcessingJob
#   include Sidekiq::Worker
#
#   def perform(data)
#     job_id = uuid_module.v4()
#     puts "Processing job #{job_id}"
#
#     # Use same UUID generator as Node.js API
#     records = data.map do |record|
#       record.merge(id: uuid_module.v4())
#     end
#
#     puts "Processed #{records.length} records"
#   end
# end

# Example 4: Validate UUIDs from API
# api_uuids = [
#   "123e4567-e89b-12d3-a456-426614174000",  # Valid
#   "invalid-uuid",                           # Invalid
#   uuid_module::NIL                          # Valid NIL
# ]
# api_uuids.each do |uuid|
#   is_valid = uuid_module.validate(uuid)
#   puts "UUID: #{uuid.ljust(45)} Valid: #{is_valid}"
# end
# puts ""

puts "Real-world use case:"
puts "- Ruby Sidekiq worker generates UUIDs for background jobs"
puts "- Uses same TypeScript implementation as Node.js API"
puts "- Guarantees consistent ID format across job queue"
puts "- No SecureRandom.uuid discrepancies"
puts ""

puts "Example: Microservices Architecture"
puts "┌─────────────────────────────────────┐"
puts "│   Elide UUID (TypeScript)          │"
puts "│   conversions/uuid/elide-uuid.ts   │"
puts "└─────────────────────────────────────┘"
puts "         ↓           ↓           ↓"
puts "    ┌────────┐  ┌────────┐  ┌────────┐"
puts "    │ Node.js│  │  Ruby  │  │ Rails  │"
puts "    │  API   │  │ Worker │  │  App   │"
puts "    └────────┘  └────────┘  └────────┘"
puts "         ↓           ↓           ↓"
puts "    Same UUID format everywhere!"
puts ""

puts "Problem Solved:"
puts "Before: Ruby SecureRandom.uuid + JavaScript uuid = different formats"
puts "After: One Elide implementation = 100% consistent UUIDs"
puts ""

puts "Rails Integration (when ready):"
puts "  # config/initializers/uuid.rb"
puts "  UUID = Elide.require('./elide-uuid.ts')"
puts "  "
puts "  # app/models/user.rb"
puts "  class User < ApplicationRecord"
puts "    before_create { self.uuid = UUID.v4() }"
puts "  end"
puts ""

puts "Sidekiq Integration (when ready):"
puts "  class MyWorker"
puts "    include Sidekiq::Worker"
puts "    "
puts "    def perform"
puts "      id = UUID.v4()"
puts "      # Process with consistent ID"
puts "    end"
puts "  end"
